text/web-servers/0003-zero-config.md

```diff
@@ -0,0 +1,75 @@
+# Support "zero-configuration" build processes
+
+## Summary
+The NGINX and HTTPD buildpacks within the Web Servers language-family should
+provide a "zero-configuration" option whereby they generate a very simple
+configuration so that the buildpack user does not need to.
+
+## Motivation
+
+Many users have very simple static file deployments. In these cases, requiring
+them to include a bunch of NGINX or HTTPD configuration boilerplate alongside
+their static files can be a burden. Instead, we should allow them to indicate
+that they have some files to serve in the simplest possible way.
+
+## Detailed Explanation
+
+The NGINX and HTTPS buildpacks will implement a common interface through an
+environment variable that will allow them to detect and build when a
+configuration file is not present in the application workspace. Specifically,
+when the `BP_WEB_SERVER` environment variable is set to either `nginx` or
+`httpd` then the respective buildpack will detect and ensure that a simple
+configuration file is included in the built image such that it can serve very
+basic static file applications.
+
+The configuration files will make assumptions about certain options, like the
+location of the files to be served. However, in some cases, the buildpacks
+could choose to allow buildpack users to configure some commonly-specified
+options (e.g. `BP_WEB_SERVER_ROOT=/workspace/my-build-directory`). This RFC
+will not enumerate these options.
+
+## Rationale and Alternatives
+
+### Why not Staticfile?
+This proposal sounds a lot like what the existing Staticfile buildpack is
+already doing, allowing users to indicate that they would like a simple
+configuration for the web server of their choice generated at build time to
+serve their static files.
+
+In the simplest cases, this is as easy as including the following
+`buildpack.yml` file in their codebase:
+
+```yaml
+staticfile:
+  nginx: {}
+```
+
+However, many users find themselves needing more than just the simplest
+configuration. They want support for SSI and HSTS, Basic Auth, and enforcement
+of HTTPS connections. This has created a slippery slope in the Staticfile
+buildpack where it is slowly recreating each of the configuration options
+already available in a proper web server configuration file with a more limited
+API. In most cases, users working with the Staticfile buildpack would likely be
+better off using the NGINX or HTTPD buildpacks directly. They'd have better
+control over their deployment configuration and the complete set of
+configuration options provided by that server.
```

Comment:
I understand the rationale here and have definitely seen users trying to escape the confines of staticfile when they should just supply their own config file. +1 for just pushing those use cases to supply their own config.

What is nice about staticfile buildpack is that there are a number of use cases that just work out-of-the-box, without needing to write config files. IMHO, the primary benefit to staticfile buildpack vs just using the nginx docker image directly is that the buildpack will generate config for you. It's not ever going to replace complicated configurations, but it's a clear advantage when you're in a hurry, like with a PoC or some short-lived project, for common and basic scenarios.

If the default config doesn't cover enough of these "it just works" use cases, there will be a gap, and users might reach for some other tool.

I'm having a little trouble determining what is turned on by the config snippet referenced above, but my $0.02 says we should cover the following scenarios:

- Hosting static files (duh - this is what it's all about)
- Push state for client-side Java apps. Not sure if this is being handled a different way, but this is the most common thing I've seen staticfile buildpack used to do.
- Hosting static files with basic auth. Basic auth isn't great, but I think there's a demand for "host my files, but don't make them public", which when paired with the next item is good enough for a number of cases.
- Force HTTPS. This is a very common requirement for larger companies. It also makes basic auth not totally broken, so it helps if you just want to throw up some files but don't want the whole world to see them.

I have seen these used very commonly with the staticfile buildpack.

Reply:
I've gone into some detail for how we could support these use cases. Take a look at the latest revision.